README.md
<div align="center">
  <h1 align="center">YouTerm</h1>

  ![GitHub releases](https://img.shields.io/github/downloads/marcoradocchia/youterm/total?color=%23a9b665&logo=github)
  ![GitHub source size](https://img.shields.io/github/languages/code-size/marcoradocchia/youterm?color=ea6962&logo=github)
  ![GitHub open issues](https://img.shields.io/github/issues-raw/marcoradocchia/youterm?color=%23d8a657&logo=github)
  ![GitHub open pull requests](https://img.shields.io/github/issues-pr-raw/marcoradocchia/youterm?color=%2389b482&logo=github)
  ![GitHub sponsors](https://img.shields.io/github/sponsors/marcoradocchia?color=%23d3869b&logo=github)
  ![PyPi downloads](https://img.shields.io/pypi/dm/youterm?color=d4be98&label=pypi%20downloads&logo=pypi&logoColor=fff)
  ![PyPi version](https://img.shields.io/pypi/v/youterm?logo=pypi&logoColor=fff)
  ![GitHub license](https://img.shields.io/github/license/marcoradocchia/youterm?color=%23e78a4e)
</div>

CLI tool to search for [YouTube](https://youtube.com) videos and play selected
video/audio via `mpv`.

![yotuerm](assets/youterm.png)

## Requirements

### YouTube API

`youterm` retrieves video informations using YouTube APIs. An API key is
needed. You can obtain one registering a Google developer profile, adding a
project and creating an API key adding the *YouTube Data API v3* service at
[this](https://console.cloud.google.com/apis/dashboard) link.

### Dependencies

`youterm` does not rely on external Python libraries, although it uses
external programs to provide audio/video streaming:

- `mpv`
- `yt-dlp` or `youtube-dl`

### Optional dependencies

To store the *YouTube* API safely the *unix standard password manager* `pass`
is required. `youterm` looks for one password entry containing the API key at
`api/youtube`, or any custom specified entry[^1], in the password-store[^2].
Alternatively you can pass the API key to `youterm` as a command line argument
(see [Usage](#Usage)) or using a config file[^3] (see
[Configuration](#Configuration)).

[^1]: You can specify custom `pass` entries for the API key using a config file
  (see [Configuration](#Configuration))
[^2]: Create entry using `pass insert api/youtube` or `pass insert <entry>` for
  custom entry (where `<entry>` is a placeholder for the actual entry name)
[^3]: *Not recommended*

## Installation

Install `youterm` using `pip`:

```bash
pip3 install youterm
```

## Usage

Invoking `youterm` with no flags defaults to audio only. Below the usage:

```
usage: youterm [-h] [-v] [-r <n>] [-a <api_key>] [-q <resolution>]

CLI tool to search for YouTube videos and play selected audio/video via MPV

options:
  -h, --help            show this help message and exit
  -v, --video           Play video
  -r <n>, --results <n>
                        Number of search results displayed
  -a <api_key>, --api <api_key>
                        YouTube Data API v3 key
  -q <resolution>, --quality <resolution>
                        Choose video quality (if not available choose closest lower)
```

## Configuration

You can specify your default options for `youterm` using a config file[^4]
located at:
[^4]: `ini` file structure

- `~/.config/youterm/config` on **Linux** and Unix based systems (following
  the XDG Base Directory specifications)
- `C:\Users\<username>\AppData\Roaming\youterm\config` on **Windows** systems
  (where `<username>` is a placeholder for the actual user name)

Below follows a configuration template:

```ini
[api]
# YouTube Data API v3 key (can be overridden using CLI argument)
key = <api_key>
# pass entry containing the API key (if you want to use pass to store the key)
# if not set defaults to: "api/youtube"
pass_entry = <pass_entry>

[search]
# default number of search results (can be overridden using CLI argument)
# if not set defaults to: 5
results = <num>

[video]
# default video resolution (can be overridden using CLI argument)
# if not set defaults to: bestvideo
# valid options are: 144, 240, 360, 480, 720, 1080, 1440, 2160
quality = <video_resolution>
```

## Changelog

For the complete changelog see [CHANGELOG.md](./CHANGELOG.md).


CHANGELOG.md
# Changelog

## [0.2.0] - 2022-05-02

### Added
- Support for configuration file
  * Added option to store the API key in the config file
  * Added option to specify a different password entry name to store the API
    key using `pass`
  * Added option to specify the default number of search results
  * Added option to specify the default video quality
- Command line argument to specify the desired **video quality** (when invoking
  `youterm` with the `-v` flag)


src/youterm/__main__.py
#!/usr/bin/env python3
# coding=utf-8

# youterm: CLI tool to search for YouTube videos and play selected video/audio
# via MPV
# Copyright (C) 2022 Marco Radocchia
#
# This program is free software: you can redistribute it and/or modify it under
# the terms of the GNU General Public License as published by the Free Software
# Foundation, either version 3 of the License, or (at your option) any later
# version.
#
# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
# details.
#
# You should have received a copy of the GNU General Public License along with
# this program. If not, see https://www.gnu.org/licenses/.

from argparse import ArgumentParser, Namespace
from configparser import ConfigParser
from os import name as os_name
from os.path import expanduser
from sys import exit
from subprocess import run
from youterm.colorizer import Colorize
from youterm.text import wrap
from youterm.yt import yt_search, get_api_key, get_details
from threading import Thread

# search history, input navigation and copy/paste (this only works on linux)
# import readline

fg = Colorize.fg
style = Colorize.style


def parse_arguments() -> Namespace:
    """
    Parse command line arguments and return argparse.Namespace
    """
    argparser = ArgumentParser(
        prog="youterm",
        description=""" CLI tool to search for YouTube videos and play selected
            video/audio via MPV""",
        allow_abbrev=False,
    )
    argparser.add_argument(
        "-v", "--video", action="store_true", help="Play video"
    )
    argparser.add_argument(
        "-r",
        "--results",
        type=int,
        metavar=("<n>"),
        help="Number of search results displayed",
    )
    argparser.add_argument(
        "-a",
        "--api",
        type=str,
        metavar=("<api_key>"),
        help="YouTube Data API v3 key",
    )
    argparser.add_argument(
        "-q",
        "--quality",
        type=str,
        metavar=("<resolution>"),
        choices=["144", "240", "360", "480", "720", "1080", "1440", "2160"],
        help="Choose video quality (if not available choose closest lower)",
    )
    return argparser.parse_args()


def parse_config() -> ConfigParser:
    """
    Parse config file and return configparser.ConfigParser
    """

    def config_file() -> str:
        if os_name == "posix":  # if os is linux/bsd/macos/...
            return expanduser("~/.config/youterm/config")
        elif os_name == "nt":  # if os is windows
            return expanduser("~\\AppData\\Roaming\\youterm\\config")
        else:
            exit("Sorry this OS is not supported")

    config = ConfigParser()
    config.read(config_file())
    return config


def get_config_option(config: ConfigParser, section: str, option: str) -> str:
    """
    Return config option if section and options appear in the config file;
    return None if not
    """
    if config.has_section(section) and config.has_option(section, option):
        return config.get(section, option)
    else:
        return None


def main_loop(api_key: str, results: int, quality: str) -> None:
    """
    Main loop
    """
    query = input("Search YouTube (`q` to quit): ")
    if query.lower() == "q":
        quit()
    try:
        response = yt_search(api_key=api_key, query=query, max_results=results)
    except Exception:
        exit(
            fg(input="ERROR: unable to get response from YouTube", color="red")
        )

    videos = []
    threads = []

    for index, item in enumerate(response["items"]):
        video = {
            "num": f"[{index+1}]",
            "id": item["id"]["videoId"],
            "channel": item["snippet"]["channelTitle"],
            "desc": item["snippet"]["description"],
        }
        threads.append(Thread(target=get_details, args=[video, api_key]))
        videos.append(video)

    # perform the requests in different threads (faster because of paralelized
    # response times)
    for thread in threads:
        thread.start()

    # wait for all threads to finish before continuing to avoid working with
    # data is not yet received
    for thread in threads:
        thread.join()

    for video in videos:
        print(
            f"├─ {fg(input=video['num'], color='red')}\n"
            "│   ├── "
            f"{wrap(input=video['title'], pref_len=8, color='yellow')}\n"
            f"│   ├── {fg(input=video['channel'], color='magenta')}\n"
            f"│   ├── {wrap(input=video['desc'], pref_len=8, color='green')}\n"
            f"│   ├── {fg(input=video['duration'], color='cyan')}\n"
            f"│   └── {fg(input=video['date'], color='blue')}"
        )

    while True:
        while True:
            try:
                selection = input(
                    "└ Select video (`q` to go back to search): "
                )
                if selection.lower() == "q":
                    return
                selection = int(selection) - 1
                if selection < 0 or selection > len(videos):
                    raise ValueError
                break
            except ValueError:
                print(
                    fg(
                        input="Please insert a valid index (integer)",
                        color="red",
                    )
                )
        selected_id = videos[selection]["id"]
        print(
            fg(input="Playing: ", color="green"),
            style(input=videos[selection]["title"], style="bolditalic"),
        )
        print(
            fg(input="URL: ", color="yellow"),
            f"https://youtu.be/{videos[selection]['id']}",
        )
        run(
            [
                "mpv",
                f"--ytdl-format={quality}",
                f"https://www.youtube.com/watch?v={selected_id}",
            ]
        )


def main() -> None:
    """
    Main function
    """
    args = parse_arguments()
    config = parse_config()

    key = (
        args.api
        or get_config_option(config=config, section="api", option="key")
        or get_api_key(
            pass_entry=get_config_option(
                config=config, section="api", option="pass_entry"
            )
        )
    )
    results = (
        args.results
        or get_config_option(config=config, section="search", option="results")
        or 5  # default value
    )
    if args.video:
        quality = args.quality or get_config_option(
            config=config, section="video", option="quality"
        )
        if quality is None:
            # no quality chosen, default to bestvideo+bestaudio:
            # "merge the best video-only format and the best audio-only format,
            # or download the best combined format if video-only format is not
            # available" (https://github.com/yt-dlp/yt-dlp#filtering-formats)
            quality = "bv+ba/b"
        else:
            quality = f"bv*[height<=?{quality}]+ba/b"
    else:
        quality = "bestaudio"

    while True:
        # enter main application loop
        main_loop(api_key=key, results=results, quality=quality)


if __name__ == "__main__":
    main()
